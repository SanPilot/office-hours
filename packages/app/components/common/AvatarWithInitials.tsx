import { Avatar } from "antd";
import React, { ReactElement } from "react";
import nameToRGB from "../../utils/ColorUtils";
import getInitialsFromName from "../../utils/NameUtils";

type AvatarWithInitalsProps = {
  name: string;
  fontSize?: number;
  style?: any;
  [x: string]: any;
};

export default function AvatarWithInitals({
  name,
  fontSize,
  style,
  ...props
}: AvatarWithInitalsProps): ReactElement {
  return (
    <Avatar
      style={{
        backgroundColor: name ? nameToRGB(name) : "#1abc9c",
        fontSize,
        ...style,
      }}
      {...props}
    >
      {getInitialsFromName(name)}
    </Avatar>
  );
}
