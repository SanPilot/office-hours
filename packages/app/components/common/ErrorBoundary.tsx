import { apm } from "@elastic/apm-rum";
import { Result } from "antd";
import React, { ReactNode } from "react";

interface EBState {
  hasError: boolean;
}
// eslint-disable-next-line @typescript-eslint/ban-types
export class ErrorBoundary extends React.Component<{}, EBState> {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error): EBState {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo): void {
    // You can also log the error to an error reporting service
    // logErrorToMyService(error, errorInfo);
    if (window) {
      apm.captureError(error);
    }
  }

  render(): ReactNode {
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return (
        <Result
          status="500"
          title="We hit an unexpected error."
          subTitle="Sorry about that! A report has automatically been filed. Try refreshing the page."
        />
      );
    }

    return this.props.children;
  }
}
