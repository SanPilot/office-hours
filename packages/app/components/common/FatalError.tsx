import Result from "antd/lib/result";
import { ReactElement } from "react";

export default function FatalError(): ReactElement {
  return (
    <Result
      status="500"
      title="We hit an unexpected error."
      subTitle="Sorry about that! A report has automatically been filed. Try refreshing the page."
    />
  );
}
