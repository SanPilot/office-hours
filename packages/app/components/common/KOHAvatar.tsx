import { UserOutlined } from "@ant-design/icons";
import { Avatar } from "antd";
import React, { ReactElement } from "react";
import { useProfile } from "../../hooks/useProfile";
import AvatarWithInitals from "./AvatarWithInitials";

type KOHAvatarProps = {
  size: number;
};

export default function KOHAvatar({ size }: KOHAvatarProps): ReactElement {
  const profile = useProfile();

  return profile?.photoURL ? (
    <Avatar
      icon={<UserOutlined />}
      src={"/api/v1/profile/get_picture/" + profile.photoURL}
      size={size}
    />
  ) : (
    <AvatarWithInitals
      style={{ marginTop: "60px", marginBottom: "60px" }}
      name={profile?.name}
      size={size}
    />
  );
}
