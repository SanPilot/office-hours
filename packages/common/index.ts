export enum WSMessageType {
  Count = "count",
  Refresh = "ref",
}

// API base data types
export type Club = { name: string; rating: number; id: number };

// API route Params and Responses
export type GetClubResponse = Club[];

export type CreateClubParams = { name: string; rating: number };
export type CreateClubResponse = Club;

export type GetCourseResponse = {
  name: string;
  officeHours: Array<{
    id: number;
    title: string;
    room: string;
    startTime: Date;
    endTime: Date;
  }>;
};
