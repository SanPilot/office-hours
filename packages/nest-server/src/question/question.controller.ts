import {
  Controller,
  Param,
  Get,
  Post,
  Body,
  HttpException,
  Patch,
} from '@nestjs/common';
import {
  CreateQuestionResponse,
  GetQuestionResponse,
  QuestionStatusKeys,
  CreateQuestionParams,
  UpdateQuestionResponse,
  UpdateQuestionParams,
} from '@template/common';
import { Queue } from 'src/queue/queue.entity';
import { Connection } from 'typeorm';
import { Question } from './question.entity';
import { User } from 'src/profile/user.entity';
import { Http2ServerResponse } from 'http2';

@Controller('questions')
export class QuestionController {
  constructor(private connection: Connection) {}

  @Get(':questionId')
  async getQuestion(
    @Param('questionId') questionId,
  ): Promise<GetQuestionResponse> {
    const question = await Question.findOne(questionId, {
      relations: ['creator', 'taHelped'],
    });

    return question;
  }

  @Post()
  async createQuestion(
    @Body() body,
  ): Promise<CreateQuestionResponse | HttpException> {
    const { text, questionType, queueId } = body as CreateQuestionParams;
    // TODO: Remove this once we implemntent user authentication
    const DEFAULT_USER = await User.create({
      id: 42,
      username: 'test_user',
      email: 'test_user@husky.neu.edu',
      name: 'Test User',
      photoURL: 'www.photoURL.com',
    }).save();
    const queueSize = await Queue.count({
      where: { id: queueId },
    });
    // Check that the queue exists
    if (queueSize === 0) {
      return new HttpException('Queue not found', 404);
    }
    // TODO: Check that the user posting the question is a member of the course

    const question = await Question.create({
      queueId: queueId,
      creator: DEFAULT_USER,
      text,
      questionType,
      status: QuestionStatusKeys.Drafting,
    }).save();

    question.creator = DEFAULT_USER;
    return question;
  }

  @Patch(':questionId')
  async updateQuestion(
    @Param('questionId') questionId,
    @Body() body,
  ): Promise<UpdateQuestionResponse | HttpException> {
    const { text, questionType } = body as UpdateQuestionParams; // Question: Do we want to take in the whole question as a param?
    // TODO: Check that the question_id belongs to the user or a TA that is currently helping with the given queue_id
    // TODO: Use user type to dertermine wether or not we should include the text in the response
    let question = await Question.findOne({
      where: { id: questionId },
      relations: ['creator'],
    });
    if (question === undefined) {
      return new HttpException('Question not found', 404);
    }
    question = Object.assign(question, body);
    await question.save();
    return question;
  }
}
