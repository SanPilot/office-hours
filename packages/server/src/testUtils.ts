import { createConnection, Connection } from "../../ormconfig";
import { init } from "./server";
import path from "path";
import { Server } from "@hapi/hapi";

export function setupServerTest(): () => Server {
  let server: Server;

  beforeEach(async () => {
    server = await init();
  });

  afterEach(async () => {
    await server.stop();
  });

  return () => server;
}

export function setupDBTest() {
  let connection: Connection;
  beforeAll(async () => {
    connection = await createConnection({
      type: "postgres",
      url: "postgres://postgres@localhost:5432/test",
      synchronize: true,
      logging: false,
      entities: [path.resolve(__dirname, "entity/**/*{.js,.ts}")],
      migrations: [path.resolve(__dirname, "migration/**/*{.js,.ts}")],
      subscribers: [path.resolve(__dirname, "subscriber/**/*{.js,.ts}")],
    });
  });

  beforeEach(async () => {
    await connection.synchronize(true);
  });

  afterAll(async () => {
    await connection.close();
  });
}
