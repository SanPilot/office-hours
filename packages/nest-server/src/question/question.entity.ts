import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
  CreateDateColumn,
} from "typeorm";
import { UserCourse } from "../profile/user-course.entity";
import { Queue } from "../queue/queue.entity";
import { QuestionType, QuestionStatus } from "@template/common";
import { User } from "../profile/user.entity";

@Entity("question_model")
export class Question extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => Queue, (q) => q.questions)
  @JoinColumn({ name: "queueId" })
  queue: Queue;

  @Column({ nullable: true })
  queueId: number;

  @Column("text")
  text: string;

  @ManyToOne((type) => User)
  @JoinColumn({ name: "creatorId" })
  creator: User;

  @Column({ nullable: true })
  creatorId: number;

  @ManyToOne((type) => User)
  @JoinColumn({ name: "taHelpedId" })
  taHelped: User;

  @Column({ nullable: true })
  taHelpedId: number;

  @CreateDateColumn()
  createdAt: Date;

  @Column({ nullable: true })
  helpedAt: Date;

  @Column({ nullable: true })
  closedAt: Date;

  @Column("text")
  questionType: QuestionType;

  @Column("text")
  status: QuestionStatus;
}
