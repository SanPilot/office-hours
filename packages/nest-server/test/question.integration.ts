import { QuestionFactory, UserFactory, QueueFactory } from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import { QuestionModule } from '../src/question/question.module';
import { Question } from '../src/question/question.entity';
import { QuestionType } from '@template/common';

describe('Question Integration', () => {
  const supertest = setupIntegrationTest(QuestionModule);

  describe('GET /questions/:id', () => {
    it('gets a question with the given id', async () => {
      const q = await QuestionFactory.create({ text: 'Help pls' });
      const response = await supertest().get(`/questions/${q.id}`).expect(200);
      expect(response.body).toMatchSnapshot();
    });
    it('fails to get a non-existent question', async () => {
      await supertest().get(`/questions/999`).expect(404);
    });
  });

  describe('POST /questions', () => {
    it('posts a new question', async () => {
      const queue = await QueueFactory.create();
      expect(await Question.count({ where: { queueId: 1 } })).toEqual(0);
      const response = await supertest()
        .post('/questions')
        .send({
          text: "Don't know recursion",
          questionType: QuestionType.Concept,
          queueId: queue.id,
        })
        .expect(201);
      expect(response.body).toMatchObject({
        text: "Don't know recursion",
        helpedAt: null,
        closedAt: null,
        questionType: 'Concept',
        status: 'Drafting',
      });
      expect(await Question.count({ where: { queueId: 1 } })).toEqual(1);
    });
    it('post question fails with non-existent queue', async () => {
      await supertest()
        .post('/questions')
        .send({
          text: "Don't know recursion",
          questionType: QuestionType.Concept,
          queueId: 999,
        })
        .expect(404);
    });
    // this test can be done once we figure out validation
    it.skip('post question fails with bad params', async () => {
      await supertest()
        .post('/questions')
        .send({
          question: 'I need help',
        })
        .expect(404);
    });
  });

  describe('PATCH /questions/:id', () => {
    it('updates a question', async () => {
      const q = await QuestionFactory.create({ text: 'Help pls' });

      const response = await supertest()
        .patch(`/questions/${q.id}`)
        .send({
          text: 'NEW TEXT',
        })
        .expect(200);
      expect(response.body).toMatchObject({
        id: q.id,
        text: 'NEW TEXT',
      });
      expect(await Question.findOne({ id: q.id })).toMatchObject({
        text: 'NEW TEXT',
      });
    });
    it('fails to update a non-existent question', async () => {
      await supertest()
        .patch(`/questions/999`)
        .send({
          text: 'NEW TEXT',
        })
        .expect(404);
    });
  });
});
