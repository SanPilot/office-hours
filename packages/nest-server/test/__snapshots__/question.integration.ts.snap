// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Integration GET /questions/:id gets a question with the given id 1`] = `
Object {
  "closedAt": null,
  "createdAt": "2020-07-05T19:38:38.102Z",
  "creator": Object {
    "email": "user1@neu.edu",
    "id": 1,
    "name": "John Doe the 1th",
    "photoURL": "https://pics/1",
    "username": "user1",
  },
  "creatorId": 1,
  "helpedAt": null,
  "id": 1,
  "questionType": "Other",
  "queueId": 1,
  "status": "Queued",
  "taHelped": null,
  "taHelpedId": null,
  "text": "Help pls",
}
`;
